File: Napakalaki-Java/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author Laura
 */
public class SpecificBadConsequence extends BadConsequence {
    private ArrayList <TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList <TreasureKind> specificVisibleTreasures = new ArrayList();
    public SpecificBadConsequence(String text_i, int lvl,  ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden){
        super(text_i,lvl);
        specificVisibleTreasures=tVisible;
        specificHiddenTreasures=tHidden;
    }
    public boolean isEmpty(){
       return (specificHiddenTreasures.isEmpty() && specificVisibleTreasures.isEmpty());
    }
    void substractVisibleTreasure(Treasure t){
       specificVisibleTreasures.remove(t.getType());
    }
    void substractHiddenTreasure(Treasure t){
       specificHiddenTreasures.remove(t.getType());
    }
    SpecificBadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
        SpecificBadConsequence out;
        ArrayList<TreasureKind> nuevoSV= new ArrayList(0);
        ArrayList<TreasureKind> copiaVT= new ArrayList(0);
        ArrayList<TreasureKind> copiaHT = new ArrayList(0); 
        ArrayList<TreasureKind> nuevoSH= new ArrayList(0);
        int index=0;
        for(Treasure treasure:v){
          copiaVT.add(treasure.getType());
        }
        for(Treasure treasure:h){
            copiaHT.add(treasure.getType());
        }
        for(TreasureKind tipo:specificVisibleTreasures){
            if((index=copiaVT.indexOf(tipo))!=-1){
                nuevoSV.add(tipo);
                copiaVT.remove(index);
            }
        }       
        for(TreasureKind tipo:specificHiddenTreasures){
            if((index=copiaHT.indexOf(tipo))!=-1){
                nuevoSH.add(tipo);
                copiaHT.remove(index);
            }
        }
        out=new SpecificBadConsequence(text,levels,nuevoSV,nuevoSH);
        System.out.println("ASDFASFDAFSSF ES EL BC DE SPECIFIC");
        System.out.println(out.toString());
        return out;
    }
    
    public ArrayList <TreasureKind> getSpecificHiddenTreasures(){
        return specificHiddenTreasures;
    }
    public ArrayList <TreasureKind> getSpecificVisibleTreasures(){
        return specificVisibleTreasures;
    }
    public String toString(){
        String out;
        out =  super.toString();
        if(specificHiddenTreasures.size()>0){
            out += ", Hidden Treasures = ";
            for (TreasureKind specificHT : specificHiddenTreasures) {
                out += ", " + specificHT.toString();
            }
        }
        if(specificVisibleTreasures!=null){
            out += ", Visible Treasures = ";
            for (TreasureKind specificVT : specificVisibleTreasures) {
                out += ", " + specificVT.toString();
            }
        }
        return out;
    }
}

File: Napakalaki-Ruby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module NapakalakiGame
class SpecificBadConsequence < BadConsequence
  def initialize(text,levels,someSpecificVisibleTreasures,someSpecificHiddenTreasures)
    super(text,levels)
    @specificVisibleTreasures=someSpecificVisibleTreasures
    @specificHiddenTreasures=someSpecificHiddenTreasures
  end
  
  def isEmpty
     (@specificHiddenTreasures.size == 0 && @specificVisibleTreasures.size == 0)
  end
  
  def substractVisibleTreasure(t)
      @specificVisibleTreasures.delete(t.getType)
  end
    
  def substractHiddenTreasure(t)
      @specificHiddenTreasures.delete(t.getType)
  end
  
  def getSpecificVisibleTreasures
    @specificVisibleTreasures
  end
  
  def getSpecificHiddenTreasures
    @specificHiddenTreasures
  end
  
  def adjustToFitTreasureLists(v,h)
    listaV=Array.new(0)
    listaH=Array.new(0)
    ajustadoV = Array.new(0)
    ajustadoH = Array.new(0)
        
    v.each do |tesoro|
      listaV << tesoro.getType
    end
    h.each do |tesoro|
      listaH << tesoro.getType
    end
        
    @specificVisibleTreasures.each do |t|
       if((indice=listaV.index(t))!=nil)
         ajustadoV << t
         listaV.delete_at(indice)
       end
    end
        
    @specificHiddenTreasures.each do |t|
      if((indice = listaH.index(t))!=nil)
        ajustadoH << t
        listaH.delete_at(indice)
      end
    end
    SpecificBadConsequence.new(@text,0,ajustadoV,ajustadoH)
  end
  
  def to_s
    out = @text+ ", pierdes #{@levels} niveles, pierdes estos tesoros visibles:\n"
    
    @specificVisibleTreasures.each do |treasure|
      out = out + treasure.to_s + " "
    end
    out= out + " y estos tesoros ocultos:\n"
    @specificHiddenTreasures.each do |treasure|
      out= out + treasure.to_s + " "
    end
    
    out     
  end
  
end
end